cli/cmd/admin/start.go

```diff
@@ -176,6 +178,16 @@ func StartCmd(cliCfg *config.Config) *cobra.Command {
 
 			// Init and run server
 			if runServer {
+				redisAddr := conf.RedisAddr
+				if redisAddr == "" {
+					// Start a miniredis (in-memory) server (is used for API rate limiting).
+					mr, err := miniredis.Run()
+					if err != nil {
+						panic(err)
+					}
+					redisAddr = mr.Addr()
+					defer mr.Close()
+				}
```

Comment:
Create the Redis client here, so it can be reused in other components in the future. If `RedisAddr` is empty, it can just be uninitialized. 

Then after, it can get a limiter by calling `ratelimit.NewRedis(redis)` if the Redis client is initialized, or `ratelimit.NewNoop()` if no Redis was initialized.

The usage of `miniredis` can then be moved to the test cases, so it doesn't get included in the production binary and can't accidentally run in production.

The init code for the above will need to be copied in the `runtime` package, but that's fine – it's just a few lines, and in return the code is decoupled :)

Reply:
good catch. I haven't noticed that `miniredis` is for testing only